### PR TITLE
[Backport 1.0] fix(hgraph): tune from available fp32 codes

### PR DESCRIPTION
**Pull request by jac0626**
## Summary

Backport #2466 to the `1.0` branch.

- Let HGraph Tune select an available lossless source from raw vectors, precise FP32 codes,
  or base FP32 codes instead of relying only on `store_raw_vector`.
- Require the selected codes to cover the current internal ID range.
- Rebuild precise codes whose payload was omitted by `ignore_reorder` serialization.
- Keep Tune state publication atomic and add source-selection, serialization, ForceRemove,
  and concurrency regression coverage.

Fixes: #2465

## Validation

```text
/opt/homebrew/opt/llvm@15/bin/clang-format --dry-run --Werror \
  src/algorithm/hgraph/hgraph.cpp tests/test_hgraph.cpp
git diff --check upstream/1.0..HEAD

source ~/.zshrc && make test CASE='[tune_codes]' COMPILE_JOBS=6
  All tests passed (8040 assertions in 1 test case)
```

## Compatibility

- API/ABI impact: none.
- This is a clean cherry-pick of the two commits from #2466.

### CODE REVIEW

File: src/algorithm/hgraph/hgraph.cpp [hunk 1]
```diff
@@ -101,8 +101,9 @@ HGraph::HGraph(const HGraphParameterPtr& hgraph_param, const vsag::IndexCommonPa
 
 bool
 HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
-    if (not this->index_feature_list_->CheckFeature(IndexFeature::SUPPORT_TUNE) or
-        not this->has_raw_vector_) {
+    std::scoped_lock lock(this->add_mutex_);
+    if (this->immutable_.load(std::memory_order_acquire) or
+        not this->index_feature_list_->CheckFeature(IndexFeature::SUPPORT_TUNE)) {
         return false;
     }
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/hgraph/hgraph.cpp:108]: Thanks for the suggestion. We checked the pre-#2466 implementation and are keeping this backport narrow. A live `HGraph` cannot contain a null or differently typed `create_param_ptr_`: the `ParamPtr` constructor casts before delegating, both the base and concrete constructors dereference the parameter during construction, and the const shared-pointer member cannot later be replaced. Moving the cast would therefore be cosmetic rather than an effective validation path, so we keep the single cast adjacent to the protected parameter update.

File: src/algorithm/hgraph/hgraph.cpp [hunk 2]
```diff
@@ -209,15 +221,24 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
     // preventing concurrent searches from accessing partially updated state.
     {
         std::scoped_lock<std::shared_mutex> wlock(this->global_mutex_);
+        auto param = std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_);
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/hgraph/hgraph.cpp:224]: This is contingent on moving the cast to the beginning, which we are not doing for the reasons in the preceding thread. The current implementation performs exactly one cast, so there is no redundant cast or variable shadowing to remove.

File: src/algorithm/hgraph/hgraph.cpp [hunk 3]
```diff
@@ -127,61 +128,72 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
...
     }
 
-    std::scoped_lock lock(this->add_mutex_);
-    if (this->immutable_.load(std::memory_order_acquire)) {
-        return false;
-    }
+    const auto current_count = total_count_.load(std::memory_order_acquire);
+    auto covers_active_ids = [current_count](const FlattenInterfacePtr& codes) {
+        return codes != nullptr and codes->TotalCount() >= current_count;
+    };
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/hgraph/hgraph.cpp:142]: We verified the actual type: `InnerIdType` is `uint32_t` in `src/basic_types.h`, and `FlattenInterface::TotalCount()` returns `InnerIdType`. The comparison is therefore unsigned `uint32_t` to unsigned `uint64_t`; it does not trigger `-Wsign-compare`, and the 1.0 build completes cleanly. An explicit cast is unnecessary.

File: src/algorithm/hgraph/hgraph.cpp [hunk 4]
```diff
@@ -127,61 +128,72 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
...
+        } catch (...) {
+            if (need_release) {
+                tune_source->Release(buffer);
+            }
+            throw;
+        }
+        if (need_release) {
+            tune_source->Release(buffer);
+        }
+    };
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/hgraph/hgraph.cpp:190]: We checked both points against the pre-#2466 path. The explicit `try`/`catch` already releases the buffer on both normal completion and exceptions. Ignoring the `Decode` boolean also preserves the existing `GetVectorByInnerId` behavior that Tune used before this refactor. Changing that failure contract or introducing a one-off scope guard would be a broader pre-existing cleanup rather than part of this backport, so we are leaving it unchanged here.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2466 to the `1.0` branch. The changes are well-structured and the test coverage is comprehensive.

**Summary of changes:**

`src/algorithm/hgraph/hgraph.cpp`:
- Moves the `add_mutex_` lock acquisition to the top of `Tune()`, replacing the early `has_raw_vector_` check with a more flexible source-selection mechanism.
- Introduces `covers_active_ids()` to validate that a candidate code source covers the current internal ID range.
- Selects a lossless FP32 tune source from raw vectors → precise codes → base codes (in priority order), instead of relying solely on `store_raw_vector`.
- Replaces the case-based logic for determining which codes need rebuilding with a cleaner boolean-driven approach.
- Consolidates `create_param_ptr_` updates inside the `global_mutex_` critical section for atomic publication.
- Adds a `decode_tune_source` helper that properly handles `GetCodesById`/`Release` lifecycle with exception safety.

`tests/test_hgraph.cpp`:
- Adds `"HGraph Tune uses available codes"` test case with 6 sections covering: raw codes, precise FP32 codes, base FP32 codes, unloaded precise codes (serialize/deserialize round-trip), force-removed codes, no usable codes fallback, and no-rebuild path.

**Review notes:**

- The existing bot comments about `static_cast<uint64_t>` and RAII/Decode-return-value have been addressed by the author with reasonable justifications — `InnerIdType` is `uint32_t` (no sign-compare issue) and the current error-handling pattern preserves pre-existing behavior.
- The remaining bot suggestions about early `param` casting are stylistic preferences; the current placement inside the `global_mutex_` critical section is functionally correct.
- Logic flow, lock ordering (`add_mutex_` → `global_mutex_`), and atomic state publication all look correct.
- Test coverage is thorough, covering the key scenarios including serialization round-trip and force-remove edge cases.

No blocking issues found. LGTM for the 1.0 backport.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2466 to the `1.0` branch. The changes are identical in logic to what was already reviewed and merged on `main`.

**Summary of changes:**
- `Tune()` now selects an available lossless FP32 source (raw vectors, precise codes, or base codes) instead of requiring `store_raw_vector`.
- `add_mutex_` acquisition is moved earlier to protect the `immutable_` check and subsequent logic consistently.
- `create_param_ptr_` updates are moved inside the `global_mutex_` critical section, making state publication atomic.
- New `decode_tune_source` lambda replaces direct `GetVectorByInnerId` calls, supporting decode from any FlattenInterface source.
- Comprehensive test coverage: separate raw codes, precise FP32 codes, base FP32 codes, unloaded precise codes after serialization, force-removed codes, no usable codes fallback, and no-rebuild path.

**Review notes:**
- The existing review comments from gemini-code-assist have been addressed by the author with reasonable justifications.
- The test coverage is thorough and covers the key edge cases (ForceRemove, serialization round-trip, fallback paths).
- No new issues identified beyond what was already discussed on #2466.

LGTM for backport.

File: src/algorithm/hgraph/hgraph.cpp (hunk 2 above)
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:224]: [suggestion] The `param` cast at line 224 is performed without a null check before dereferencing it on subsequent lines. Adding a null check would be consistent with defensive coding practices.

**Review by LHT129 — COMMENTED** [review]
## Review Summary

This is a clean backport of #2466 to the `1.0` branch. The changes allow HGraph Tune to select an available lossless FP32 source (raw vectors, precise codes, or base codes) instead of relying solely on `store_raw_vector`.

### What was reviewed

- **`src/algorithm/hgraph/hgraph.cpp`** (+60/-39): The core logic refactor in `Tune()` — lock ordering change, `covers_active_ids` helper, `tune_source` cascade selection, `decode_tune_source` lambda, and atomic state publication under `global_mutex_`.
- **`tests/test_hgraph.cpp`** (+111): Seven new test sections covering separate raw codes, precise FP32 codes, base FP32 codes, unloaded precise codes after serialization, force-removed codes, no usable codes, and no-rebuild scenarios.

### Assessment

The diff is well-structured and the logic is sound. The cascade selection (`raw_vector_` → `high_precise_codes_` → `basic_flatten_codes_`) correctly requires FP32 quantization for the code-based sources and validates coverage via `covers_active_ids`. Moving `create_param_ptr_` mutation inside the `global_mutex_` critical section fixes the atomicity issue noted in the original PR. Test coverage is thorough, including serialization round-trip and force-remove edge cases.

Existing review comments from other reviewers have already addressed the notable suggestions (null check on `param` cast, signed/unsigned comparison in `covers_active_ids`). No additional blocking issues found.

**Review by LHT129 — COMMENTED** [review]
Review summary for PR #2475:

This is a clean backport of #2466 to the `1.0` branch. The changes are well-structured and the test coverage is comprehensive.

**One remaining concern:**

`src/algorithm/hgraph/hgraph.cpp:224` — `std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_)` is used without a null check before dereferencing `param` on subsequent lines. While `create_param_ptr_` is expected to always be an `HGraphParameter` in practice, adding a null check would be consistent with defensive coding practices.

Otherwise LGTM for the backport.

File: src/algorithm/hgraph/hgraph.cpp (hunk 2 above)
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:224]: [suggestion] `std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_)` can return `nullptr` if `create_param_ptr_` is not actually an `HGraphParameter`. The subsequent lines (e.g., `param->precise_codes_param.reset()`, `param->base_codes_param`, `param->use_reorder`, `param->reorder_source`) dereference `param` unconditionally, which would be undefined behavior. Consider adding a null check: `if (param == nullptr) { return false; }`

File: src/algorithm/hgraph/hgraph.cpp [hunk 5]
```diff
@@ -127,61 +128,72 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
...
-        new_use_reorder = inner_parameter->use_reorder;
-        drop_precise_codes = true;
-        param->precise_codes_param.reset();
-        is_tune_precise_code = false;
-    }
-    if (not new_use_reorder and inner_parameter->use_reorder and not new_reorder_by_base) {
-        // [case 4] assign new precise_code
-        new_use_reorder = true;
+    if (need_precise_codes and
+        (not covers_active_ids(high_precise_codes_) or
```
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:152]: [critical] Null pointer dereference when `high_precise_codes_` is null but `need_precise_codes` is true.

At line 151-154:
```cpp
if (need_precise_codes and
    (not covers_active_ids(high_precise_codes_) or
     high_precise_codes_->GetQuantizerName() != new_precise_code->GetQuantizerName())) {
```

When `high_precise_codes_` is null, `covers_active_ids` returns false (due to the `codes != nullptr` guard). The `not` inverts it to true, and because `or` short-circuits on the first true operand, the second condition `high_precise_codes_->GetQuantizerName()` is **not** short-circuited — it is evaluated, causing a null pointer dereference.

This scenario can occur when `need_precise_codes` is true (i.e., `use_reorder && !reorder_by_base`) but the index was built without precise codes (e.g., serialized with `ignore_reorder`).

Suggested fix — add an explicit null check before accessing `high_precise_codes_`:
```cpp
if (need_precise_codes and
    (not covers_active_ids(high_precise_codes_) or
     (high_precise_codes_ != nullptr and
      high_precise_codes_->GetQuantizerName() != new_precise_code->GetQuantizerName()))) {
```

Alternatively, restructure to avoid the double evaluation:
```cpp
if (need_precise_codes and not covers_active_ids(high_precise_codes_)) {
    is_tune_precise_code = true;
} else if (need_precise_codes and
           high_precise_codes_->GetQuantizerName() != new_precise_code->GetQuantizerName()) {
    is_tune_precise_code = true;
}
```

**Review by LHT129 — COMMENTED** [review]
I have reviewed this backport of #2466 to the `1.0` branch. The existing review comments from LHT129 and gemini-code-assist have already covered the key issues:

1. **[critical] Line 151-153**: Null pointer dereference when `high_precise_codes_` is null but `need_precise_codes` is true — the `or` short-circuit does not protect the second operand `high_precise_codes_->GetQuantizerName()`. This needs to be fixed before merging.

2. **[critical] Line 224**: `std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_)` result is dereferenced without a null check at lines 228, 232, 236, 240, 241. Add a null guard.

3. **[suggestion] Line 141**: `codes->TotalCount()` returns a signed type compared against `uint64_t current_count`, which may trigger `-Wsign-compare`. Add a `static_cast<uint64_t>`.

4. **[suggestion] Line 180**: `Decode` return value is unchecked. Consider verifying the return value and throwing on failure.

5. **[suggestion] Lines 172-190**: The manual try-catch for `Release` cleanup could be simplified with an RAII scope guard.

The test coverage is comprehensive, covering raw codes, precise codes, base codes, serialization round-trip with `ignore_reorder`, force-remove, no-usable-codes, and no-rebuild scenarios. The test structure and assertions look correct.

Please address the two critical null-pointer issues before merging.

**Review by LHT129 — COMMENTED** [review]
I have reviewed this backport of #2466 to the `1.0` branch. The changes look clean and the cherry-pick appears faithful to the original.

The existing inline review comments from other reviewers have already covered the key points:
- Null check for `param` after `dynamic_pointer_cast` (line 224)
- Sign-compare warning for `TotalCount()` vs `uint64_t` (line 142)
- `Decode` return value should be checked in `decode_tune_source` (line 172-190)
- RAII-based cleanup in `decode_tune_source` (line 172-190)

Regarding the [critical] comment about null pointer dereference at line 152: C++ short-circuit evaluation with `or` means that when `high_precise_codes_` is null, `covers_active_ids` returns false, `not false` = true, and the right operand `high_precise_codes_->GetQuantizerName()` is never evaluated. The code is correct, though an explicit null check would improve readability.

The test coverage is comprehensive, covering raw codes, precise codes, base codes, ignore_reorder serialization, force-remove, no-usable-codes, and no-rebuild scenarios.

No additional blocking issues found.

File: src/algorithm/hgraph/hgraph.cpp (hunk 5 above)
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:152]: [note] Regarding the null pointer concern at line 151-154:

```cpp
if (need_precise_codes and
    (not covers_active_ids(high_precise_codes_) or
     high_precise_codes_->GetQuantizerName() != new_precise_code->GetQuantizerName())) {
```

When `high_precise_codes_` is null, `covers_active_ids` returns `false` (due to the `codes != nullptr` guard). `not false` evaluates to `true`. Since C++ `or` short-circuits when the left operand is `true`, the right operand `high_precise_codes_->GetQuantizerName()` is **never evaluated** in this case. Therefore, there is no null pointer dereference here — the short-circuit evaluation correctly protects against it.

**Review by LHT129 — COMMENTED** [review]
Thank you for this backport. The changes look well-structured and the test coverage is comprehensive, covering multiple tune-source scenarios (raw, precise FP32, base FP32, unloaded precise codes, force-remove, no usable codes, and no-rebuild).

I have reviewed the diff and the existing inline comments already cover the main points worth addressing:

1. **`Decode` return value**: The `Decode` call in `decode_tune_source` returns a `bool` that is currently ignored. Checking it would prevent silent failures during training/rebuild.

2. **Exception safety in `decode_tune_source`**: The manual try-catch with duplicated `Release` calls could be simplified with a scope guard, improving readability and guaranteeing cleanup.

3. **Sign-compare in `covers_active_ids`**: `TotalCount()` returns a signed type compared against `uint64_t`, which may trigger `-Wsign-compare` warnings.

4. **`param` cast placement**: Moving the `dynamic_pointer_cast<HGraphParameter>` to the top of `Tune` would allow early validation and avoid the redundant cast later.

The null-pointer concern about `high_precise_codes_` at line 151-154 has been clarified — C++ `or` short-circuit evaluation correctly protects against dereferencing a null `high_precise_codes_` when `covers_active_ids` returns false.

The test additions are thorough and cover the key scenarios well. No blocking issues from my side.

**Review by LHT129 — COMMENTED** [review]
I have reviewed this PR and the existing review comments already cover the key areas of concern:

- The gemini-code-assist bot has raised points about early `param` casting, sign-compare in `covers_active_ids`, and `Decode` return-value checking + RAII in `decode_tune_source`. The author has responded to each with reasoning for keeping the backport narrow.
- LHT129 has raised a defensive null-check suggestion for the `param` cast at the critical section entry, and also raised (then retracted) a concern about `high_precise_codes_` null dereference which is correctly protected by short-circuit evaluation.

The diff logic is sound: the tune-source selection fallback chain (raw → precise fp32 → base fp32) is correct, the `add_mutex_` lock is properly moved earlier to guard the early-return checks, and the parameter updates are now performed atomically under `global_mutex_`. The test coverage is comprehensive, covering all code paths including the force-remove and ignore_reorder serialization edge cases.

I have no additional blocking concerns to raise beyond what is already discussed in the existing threads.

**Review by LHT129 — COMMENTED** [review]
## Review Summary

This is a clean backport of #2466 to the `1.0` branch. The changes allow HGraph Tune to select an available lossless FP32 source (raw vectors, precise codes, or base codes) instead of relying only on `store_raw_vector`.

### What was reviewed

- **`src/algorithm/hgraph/hgraph.cpp`** (+60/-39): Refactored `Tune()` to select tune source from available FP32 codes, moved lock acquisition earlier, added `covers_active_ids` helper, and consolidated state update under the global mutex.
- **`tests/test_hgraph.cpp`** (+111/-0): Added comprehensive `"HGraph Tune uses available codes"` test with 7 scenarios covering separate raw codes, precise FP32 codes, base FP32 codes, unloaded precise codes after serialization, force-removed codes, no usable codes, and no-rebuild paths.

### Key observations

1. **Null pointer safety at line 151-154**: The `or` short-circuit evaluation correctly protects against null `high_precise_codes_` dereference — when `covers_active_ids` returns false for a null pointer, `not false` is true and the right-hand `GetQuantizerName()` call is never evaluated. The earlier concern raised by LHT129 was correctly resolved.

2. **`Decode` return value not checked** (line 181): `tune_source->Decode(buffer, data)` returns a `bool` indicating success/failure. Ignoring this return value means decode failures could silently corrupt training/rebuild data. This was previously noted by gemini-code-assist and remains unaddressed.

3. **Sign-compare in `covers_active_ids`** (line 141): `codes->TotalCount()` returns `InnerIdType` (signed) while `current_count` is `uint64_t`. A `static_cast<uint64_t>` would avoid `-Wsign-compare` warnings, consistent with the project's preference for `uint64_t` over `size_t`. Also previously noted by gemini-code-assist.

### Test coverage

The new test covers all the key scenarios described in the PR summary, including the tricky `ignore_reorder` serialization/deserialization path and the force-remove edge case. The test structure is clean and well-organized.

### Overall

The backport is faithful to the original #2466 changes. The existing review comments from gemini-code-assist and LHT129 have already covered the main improvement opportunities. The two remaining unaddressed items (Decode return value check, sign-compare cast) are suggestions, not blockers.

File: src/algorithm/hgraph/hgraph.cpp [hunk 6]
```diff
@@ -127,61 +128,72 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
...
+            throw VsagException(ErrorType::INTERNAL_ERROR,
+                                fmt::format("failed to get vector by inner id {}", inner_id));
+        }
+        try {
+            tune_source->Decode(buffer, data);
+        } catch (...) {
+            if (need_release) {
+                tune_source->Release(buffer);
+            }
+            throw;
```
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:185]: [suggestion] The return value of `tune_source->Decode(buffer, data)` is ignored. If `Decode` returns `false` (indicating a decode failure), the training/rebuild data will be silently corrupted with uninitialized or stale float values in the output buffer.

This is a follow-up to the earlier gemini-code-assist suggestion which also recommended checking the return value.

Suggested fix:
```cpp
if (not tune_source->Decode(buffer, data)) {
    throw VsagException(ErrorType::INTERNAL_ERROR,
                        fmt::format("failed to decode vector by inner id {}", inner_id));
}
```

File: src/algorithm/hgraph/hgraph.cpp (hunk 2 above)
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:224]: [critical] `dynamic_pointer_cast` at line 224 may return `nullptr` if `create_param_ptr_` is not of type `HGraphParameter`. The returned pointer `param` is dereferenced unconditionally on lines 228, 232, 236, 240, 241, and 243 without any null check, which would cause a null pointer dereference and crash.

Suggested fix: add a null check after the cast and return `false` (or throw) if `param` is null.

```cpp
auto param = std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_);
if (param == nullptr) {
    return false;
}
```

File: src/algorithm/hgraph/hgraph.cpp [hunk 7]
```diff
@@ -209,15 +221,24 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
...
     {
         std::scoped_lock<std::shared_mutex> wlock(this->global_mutex_);
+        auto param = std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_);
         basic_flatten_codes_ = new_basic;
         if (drop_precise_codes) {
             high_precise_codes_.reset();
+            param->precise_codes_param.reset();
         } else {
             high_precise_codes_ = new_precise;
+            if (is_tune_precise_code) {
```
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:231]: [critical] When `need_precise_codes` is true but `is_tune_precise_code` is false (i.e., the existing precise codes already have the correct quantizer and cover all active IDs), `new_precise` is set to the untrained `new_precise_code` via `rebuild_code(new_precise_code, false)`. This untrained instance then replaces the existing trained `high_precise_codes_` at line 231, effectively discarding all trained precise codes.

The same issue applies to `basic_flatten_codes_` at line 226 when `is_tune_base_code` is false — the untrained `new_basic_code` replaces the trained `basic_flatten_codes_`.

The `rebuild_code` lambda returns the untrained `new_code` when `need_rebuild` is false, but the caller unconditionally assigns the result. The state update should skip replacement when no rebuild was performed.

Suggested fix:
```cpp
basic_flatten_codes_ = new_basic;
if (drop_precise_codes) {
    high_precise_codes_.reset();
    param->precise_codes_param.reset();
} else if (is_tune_precise_code) {
    high_precise_codes_ = new_precise;
    param->precise_codes_param = hgraph_parameter->precise_codes_param;
}
```

This ensures `high_precise_codes_` is only replaced when it was actually rebuilt.

File: src/algorithm/hgraph/hgraph.cpp [hunk 8]
```diff
@@ -127,61 +128,72 @@ HGraph::Tune(const std::string& parameters, bool disable_future_tuning) {
...
+            tune_source = basic_flatten_codes_;
+        } else {
+            return false;
+        }
     }
-    param->use_reorder = new_use_reorder;
-    param->reorder_source = inner_parameter->reorder_source;
 
-    // export train data and train new_basic_code
+    auto decode_tune_source = [&](InnerIdType inner_id, float* data) {
```
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:172]: [suggestion] The `decode_tune_source` lambda is defined even when `tune_source` is null (i.e., when neither `is_tune_base_code` nor `is_tune_precise_code` is true). While it is only invoked inside guarded `if (is_tune_base_code or is_tune_precise_code)` blocks, defining a lambda that would dereference a null pointer if called is fragile and could lead to issues during future refactoring.

Consider moving the lambda definition inside the `if (is_tune_base_code or is_tune_precise_code)` block, or restructuring so `tune_source` is always valid when the lambda exists.

**Review by LHT129 — COMMENTED** [review]
[note] This is a clean backport of #2466 to the `1.0` branch. The diff is identical in substance to the original PR.

**Summary of existing review coverage:**

The PR has already been thoroughly reviewed with 16 inline comments:
- **LHT129** flagged 8 comments covering: null safety on `dynamic_pointer_cast` (line 224), short-circuit safety of `covers_active_ids` with null `high_precise_codes_` (line 152), ignored `Decode` return value (line 185), untrained codes swap concern (line 231), and lambda scope (line 172).
- **gemini-code-assist** flagged 4 medium-priority items (lines 108, 142, 190, 224), all addressed by the author.

**What this PR does well:**
- Moves the `add_mutex_` acquisition to the beginning of `Tune()`, preventing TOCTOU on `immutable_` check.
- Introduces a systematic source-selection cascade (raw_vector → precise FP32 → base FP32) with `covers_active_ids` guard.
- Properly handles `ignore_reorder` serialization edge case by rebuilding precise codes from available FP32 sources.
- Atomic state publication under `global_mutex_` exclusive lock.
- Comprehensive test coverage: 7 test sections covering raw codes, precise FP32, base FP32, unloaded precise codes (serialize/deserialize round-trip), force-remove, no-usable-codes, and no-rebuild paths.

**No new blocking issues found.** The remaining open LHT129 comments are acknowledged suggestions/notes. This backport is ready to merge once the existing review threads are resolved.

**Review by LHT129 — COMMENTED** [review]
## Review Summary

This is a backport of #2466 to the `1.0` branch. The change allows HGraph Tune to select an available lossless FP32 source (raw vectors, precise codes, or base codes) instead of relying solely on `store_raw_vector`.

### Existing Review Coverage

The PR already has several substantive review comments that I agree with and will not duplicate:

1. **[critical] by @LHT129** — When `is_tune_base_code` or `is_tune_precise_code` is false, the `rebuild_code` lambda returns the untrained code instance, which then unconditionally replaces the existing trained codes in the state update block. This is a correctness bug that would discard trained quantization data.

2. **[suggestion] by @LHT129** — The `decode_tune_source` lambda is defined even when `tune_source` is null, which is fragile for future refactoring.

3. **[suggestion] by @gemini-code-assist** — Several improvements: moving `param` cast to function entry with nullptr check, sign-compare warning fix for `TotalCount()` comparison, and exception-safety / unchecked-Decode-return in `decode_tune_source`.

### Test Coverage

The new test case `"HGraph Tune uses available codes"` provides good coverage of the new code paths: raw vector fallback, precise FP32 codes, base FP32 codes, deserialized codes with `ignore_reorder`, force-removed entries, no-usable-codes, and no-rebuild scenarios. This is thorough for a backport.

### Recommendation

The critical issue identified by @LHT129 should be addressed before merging. The state update block should only replace `basic_flatten_codes_` / `high_precise_codes_` when the corresponding rebuild was actually performed.

**Review by LHT129 — COMMENTED** [review]
Thanks for this backport. The existing reviews from @gemini-code-assist[bot] and @LHT129 have already covered the key issues comprehensively:

1. **[critical] Untrained codes replacing trained ones** (@LHT129): When `is_tune_base_code` or `is_tune_precise_code` is false, the `rebuild_code` lambda returns an untrained instance that then replaces the existing trained codes. This affects both `basic_flatten_codes_` (line 226) and `high_precise_codes_` (line 231).

2. **[suggestion] `decode_tune_source` improvements** (@gemini-code-assist): Consider RAII for resource cleanup and check `Decode()` return value.

3. **[suggestion] Sign-compare warning** (@gemini-code-assist): `TotalCount()` returns a signed type compared against `uint64_t current_count`.

4. **[suggestion] Lambda defined with null `tune_source`** (@LHT129): `decode_tune_source` is defined even when `tune_source` is null.

The critical issue #1 should be addressed before merging. The test coverage looks good with 7 sections covering the key scenarios.

File: src/algorithm/hgraph/hgraph.cpp (hunk 7 above)
Comment (LHT129) [reply src/algorithm/hgraph/hgraph.cpp:231]: [note] This analysis is incorrect — the `tune_and_rebuild` lambda returns `old_code` when `need_tune` is false, not the untrained `new_code`:

```cpp
auto tune_and_rebuild =
    [&](bool need_tune, FlattenInterfacePtr old_code, FlattenInterfacePtr new_code) {
        if (not need_tune) {
            return old_code;  // returns existing trained codes
        }
        // ... train and rebuild ...
        return new_code;
    };
```

So when `is_tune_precise_code` is false, `new_precise = tune_and_rebuild(false, high_precise_codes_, new_precise_code)` returns `high_precise_codes_`, and `high_precise_codes_ = new_precise` is a self-assignment (no-op). The same applies to `basic_flatten_codes_`. There is no correctness bug here.

The only minor concern is that `new_precise_code` is allocated via `FlattenInterface::MakeInstance` even when `is_tune_precise_code` is false (i.e., when it will never be used), which is a wasted allocation but not a correctness issue.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2466 to the `1.0` branch. The diff matches the original PR closely.

**Summary of existing review discussions (all resolved or acknowledged):**

1. **`param` cast placement** (gemini-code-assist): Suggested moving the `dynamic_pointer_cast` to the function entry for early validation. Author declined — `create_param_ptr_` is guaranteed to be `HGraphParameter` by construction, making the early cast cosmetic rather than an effective validation path. The single cast adjacent to the parameter update is sufficient.

2. **Sign-compare in `covers_active_ids`** (gemini-code-assist): Suggested casting `TotalCount()` to `uint64_t`. Author verified that `InnerIdType` is `uint32_t`, so the comparison is unsigned-to-unsigned and does not trigger `-Wsign-compare`.

3. **`decode_tune_source` improvements** (gemini-code-assist, LHT129): Suggestions for RAII scope guard, `Decode` return value check, and moving the lambda inside the guard block. Author declined — these preserve the pre-#2466 behavior and changing the failure contract would be a broader cleanup beyond this backport.

4. **Null dereference in precise codes check** (LHT129): Initially flagged as [critical], then self-corrected — C++ `or` short-circuit evaluation correctly protects against null dereference when `high_precise_codes_` is null.

5. **Untrained codes replacing trained ones** (LHT129): Initially flagged as [critical], then self-corrected — `tune_and_rebuild` returns `old_code` when `need_tune` is false, so the assignment is a self-assignment (no-op).

**Test coverage**: The new `TEST_CASE("HGraph Tune uses available codes")` covers all key scenarios: raw vector source, precise FP32 source, base FP32 source, unloaded precise codes after serialization, force-remove compatibility, no-usable-codes fallback, and no-rebuild path.

This backport looks correct and complete. No blocking issues remain.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2466 to the `1.0` branch. The changes look correct and the test coverage is comprehensive, covering all tune source selection paths (raw vectors, precise FP32 codes, base FP32 codes, unloaded precise codes after serialization, force-removed scenarios, no-usable-codes fallback, and no-rebuild transitions).

The existing review comments from gemini-code-assist and LHT129 have already addressed the notable items (early `param` validation, sign-compare in `covers_active_ids`, `Decode` return value checking, and lambda placement). No additional blocking issues found.

LGTM for the backport.

File: src/algorithm/hgraph/hgraph.cpp (hunk 2 above)
Comment (LHT129) [inline src/algorithm/hgraph/hgraph.cpp:224]: [suggestion] The `param` pointer obtained via `std::dynamic_pointer_cast<HGraphParameter>(create_param_ptr_)` inside the `global_mutex_` critical section is dereferenced without a null check (e.g., `param->precise_codes_param.reset()`, `param->base_codes_param = ...`, `param->use_reorder = ...`). While `create_param_ptr_` is expected to always hold a valid `HGraphParameter` in practice, a null guard would make this code more defensive.

Suggested fix: add a null check after the cast, or hoist the cast to function entry and validate once.
